### PR TITLE
Allow properties to hold a list of interested clients

### LINKED ISSUE #95: As a property agent, I can add properties along with interested clients

### LINKED ISSUE #96: As a property agent, I can edit the interested client list of properties

### PR DESCRIPTION
**Pull request by lkwlkww**
close #95 close #96 

Previously, properties and clients were quite segregated in Condonery. Properties can now hold reference to a list of interested clients.

## Features
1. When adding properties, the user can choose to add a list of interested clients via the `ic/` tag 
  - Example: `add -p n/randomname a/randomadd t/tag1 tag2 ic/clientname1 clientname2`
2. When editing properties, the user can edit the list interested clients
3. The user can use the `select -p [INDEX]` command to view the details of a property, as well as display its list of interested clients.

## Todo
1. Update GUI to reflect a property's interested clients. (Maybe can put no. of interested clients, and using the `select` command would show the detailed view)
2. Make it clearer to the user that what is displayed is an interested client list, after a `select` command has been carried out
3. Update test files to fit all changes and to pass CI
4. Add in defensive programming attributes

# Screenshot
Screenshot 2022-10-21 at 3 10 23 PM

### CODE REVIEW

File: src/main/java/seedu/condonery/logic/parser/ParserUtil.java [hunk 1]
```diff
@@ -89,4 +90,34 @@ public static Set<Tag> parseTags(Collection<String> tags) throws ParseException
...
+     * Parses a {@code String clientName} into a {@code Client}.
+     * Leading and trailing whitespaces will be trimmed.
+     *
+     * @throws ParseException if the given {@code clientName} does not match to a unique
+     * existing client's name.
+     */
+    public static Client parseClientName(String clientName) throws ParseException {
+        requireNonNull(clientName);
+        String trimmedClientName = clientName.trim();
+        return new Client(
```
Comment (misterpuffin) [inline src/main/java/seedu/condonery/logic/parser/ParserUtil.java:104]: I'm not very sure if this function should return a 'Client'. I understand that this is being stored in the Property as interestedClients, but it's abit strange to create clients with substring names and placeholder addresses, I think I prefer if it stores the actual client instead.

I see that in the AddPropertyCommand you use the parser to find the actual references to the clients. Is it possible to just store these references to the clients directly in the Property?

**Review by misterpuffin — APPROVED** [review]
am ok with the changes for now! but i think might be to store the actual client instead of the substring in the property model/storage also.

File: src/main/java/seedu/condonery/logic/parser/ParserUtil.java (hunk 1 above)
Comment (lkwlkww) [reply src/main/java/seedu/condonery/logic/parser/ParserUtil.java:104]: Thanks hr! I agree that it's kinda weird. There were some issues I faced and I thought I would just implement this in the simplest way possible first, then improve it in future iterations.
I haven't figured out how to solve those issues, but perhaps I could ask you for suggestions if you don't mind?

First, some considerations:
- The `ModelManager` object is required (e.g. AddPropertyCommand takes in a `Model` object in its `execute()` function) for finding matches to `Clients`, based on a user's input, since a `ModelManager` is initialised at the opening of Condonery, and a `ModelManager` holds reference to the most updated `ClientDirectory`
- Among all the involved classes in a typical command call, only the `AddPropertyCommand` class interacts with a `ModelManager` object. Thus, the matching of `Clients` to user input has to be done in the `AddPropertyCommand ` class


Basically, I have considered storing those references to the clients directly in the `Property`, but thought it would be a bit weird because with this then a `Property` would contain a
```java
Set<Client> interestedClients
```
as well as a
```java
Set<String> interestedClientNames // names found in the user's argument
```

I thought it would be weird because then Properties would directly hold reference to a user's input, while

Comment (lkwlkww) [reply src/main/java/seedu/condonery/logic/parser/ParserUtil.java:104]: Thanks hr! I agree that it's kinda weird. There were some issues I faced and I thought I would just implement this in the simplest way possible first, then improve it in future iterations.
I haven't figured out how to solve those issues, but perhaps I could ask you for suggestions if you don't mind?

First, some considerations:
- The `ModelManager` object is required (e.g. AddPropertyCommand takes in a `Model` object in its `execute()` function) for finding matches to `Clients`, based on a user's input, since a `ModelManager` is initialised at the opening of Condonery, and a `ModelManager` holds reference to the most updated `ClientDirectory`
- Among all the involved classes in a typical command call, only the `AddPropertyCommand` class interacts with a `ModelManager` object. Thus, the matching of `Clients` to user input has to be done in the `AddPropertyCommand ` class
- Currently, I can only access the `ClientDirectory` in the `AddPropertyCommand` class. I decided to create a `Property` with the placeholder `interestedClients` first, before passing it into `AddPropertyCommand`. Thus, in `AddPropertyCommand`, I can access the user's input via the placeholder `interestedClients` in the `Property`, and also access the `ClientDirectory` through the `Model` object.

The issue is that I could not think of a better way to design this, as certain classes don't normally have access to certain data. 

One way to change this I think, is to make a new method in a `Property`: `addInterestedClients`, while the `AddPropertyCommand` would take in a new Property as well as the user's input. Then, the `AddPropertyCommand` would match `Clients` to the user input as usual before passing those `Clients` into the `addInterestedClients` of the new `Property`.

I hope I have been clear, let me know if there are parts that require more clarification. Overall, WDYT?

**Review by zacchaeuschok — COMMENTED** [review]
- will you be adding an edit feature for interested clients as well?

**Review by zacchaeuschok — APPROVED** [review]

File: src/main/java/seedu/condonery/logic/parser/ParserUtil.java (hunk 1 above)
Comment (misterpuffin) [reply src/main/java/seedu/condonery/logic/parser/ParserUtil.java:128]: ah i see what you mean! I was thinking instead of storing the `interestedClients` placeholder within the Property first, maybe we can store the `Set<String>` into the `AddPropertyCommand` class first. The reason being as currently it implies that the `Client` objects in Property are the real clients, and I'm afraid in the future we forget to update the references it stores. But by keeping the queries as strings, we avoid this confusion. 

Then when the `AddPropertyCommand` executes, we can do the same logic of querying the `Model`. I agree with the suggestion of `addInterestedClients`, and we can call this method before adding the `Property` to the directory. 

But yup! I think code quality wise it is okay to improve in the future! good to make sure the feature works first before we try to refactor it.

**Comment by lkwlkww** [comment]
close #112 

With the latest commit, `Clients` are now also able to hold reference to a list of `properties` that the client is interested in.

File: src/main/java/seedu/condonery/logic/commands/property/AddPropertyCommand.java [hunk 2]
```diff
@@ -5,11 +5,18 @@
...
+import java.util.Set;
+import java.util.stream.Collectors;
+
 import seedu.condonery.logic.commands.Command;
 import seedu.condonery.logic.commands.CommandResult;
 import seedu.condonery.logic.commands.exceptions.CommandException;
 import seedu.condonery.model.Model;
 import seedu.condonery.model.property.Property;
+import seedu.condonery.model.property.utils.ParsePropertyInterestedClients;
+import seedu.condonery.model.client.Client;
```
[the remark this answers is not in the document]
Comment (lkwlkww) [reply src/main/java/seedu/condonery/logic/commands/property/AddPropertyCommand.java:19]: Good catch, thanks! I think we should.
